datastore/statedata.go

```
@@ -0,0 +1,267 @@
+	MenderStateDone
+)
+
+var (
+	stateNames = map[MenderState]string{
+		MenderStateInit:                   "init",
+		MenderStateIdle:                   "idle",
+		MenderStateAuthorize:              "authorize",
+		MenderStateAuthorizeWait:          "authorize-wait",
+		MenderStateInventoryUpdate:        "inventory-update",
+		MenderStateCheckWait:              "check-wait",
+		MenderStateUpdateCheck:            "update-check",
+		MenderStateUpdateFetch:            "update-fetch",
+		MenderStateUpdateStore:            "update-store",
+		MenderStateUpdateAfterStore:       "update-after-store",
+		MenderStateUpdateInstall:          "update-install",
+		MenderStateFetchStoreRetryWait:    "fetch-install-retry-wait",
+		MenderStateUpdateVerify:           "update-verify",
+		MenderStateUpdateCommit:           "update-commit",
+		MenderStateUpdateAfterFirstCommit: "update-after-first-commit",
+		MenderStateUpdateAfterCommit:      "update-after-commit",
+		MenderStateUpdateStatusReport:     "update-status-report",
+		MenderStatusReportRetryState:      "update-retry-report",
+		MenderStateReportStatusError:      "status-report-error",
+		MenderStateReboot:                 "reboot",
+		MenderStateVerifyReboot:           "verify-reboot",
+		MenderStateAfterReboot:            "after-reboot",
+		MenderStateRollback:               "rollback",
+		MenderStateRollbackReboot:         "rollback-reboot",
+		MenderStateVerifyRollbackReboot:   "verify-rollback-reboot",
+		MenderStateAfterRollbackReboot:    "after-rollback-reboot",
+		MenderStateError:                  "error",
+		MenderStateUpdateError:            "update-error",
+		MenderStateUpdateCleanup:          "cleanup",
+		MenderStateDone:                   "finished",
+	}
+)
+
+func (m MenderState) MarshalJSON() ([]byte, error) {
+	n, ok := stateNames[m]
+	if !ok {
+		return nil, fmt.Errorf("marshal error; unknown state %v", m)
+	}
+	return json.Marshal(n)
+}
+
+func (m MenderState) String() string {
+	return stateNames[m]
```

Shouldn't  we check here also that stateNames[m] exist? Same as in MarshalJSON

---

Hmm, this is a tricky case. The `String()` is primarily implemented so that `Marshal()` will use the string representation instead of the int, to store the state in the database. It cannot have an error return, so our only other option is to store something like `"unknown"`. But this is not very good, because it would leave a permanent incorrect record in the database. It's not supposed to happen unless there is a programming error somewhere, so I think I would actually vote for getting a panic here instead of silently putting a broken record in the database. At least that way we'll know.

---

Understood, thanks :+1: